File: src/main/java/pers/neige/neigeitems/action/node/impl/JoinParser.java
package pers.neige.neigeitems.action.node.impl;

import lombok.NonNull;
import lombok.val;
import lombok.var;
import org.bukkit.OfflinePlayer;
import org.jetbrains.annotations.Nullable;
import pers.neige.neigeitems.action.ActionContext;
import pers.neige.neigeitems.action.node.NodeParser;
import pers.neige.neigeitems.config.ConfigReader;
import pers.neige.neigeitems.manager.BaseActionManager;
import pers.neige.neigeitems.script.CompiledScript;
import pers.neige.neigeitems.utils.NumberParser;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ThreadLocalRandom;
import java.util.function.Function;

/**
 * 对齐旧 section JoinParser：支持 transform / 空列表拼前后缀 /
 * 非法 limit 视为无限制 / 非法 shuffled 视为 false / 先 shuffle 再逐项解析。
 */
public class JoinParser extends NodeParser {
    /**
     * 用于 join 节点的已编译 transform 脚本缓存（与旧 section 一致）。
     */
    private static final ConcurrentHashMap<String, CompiledScript> compiledScripts = new ConcurrentHashMap<>();

    public JoinParser(@NonNull BaseActionManager manager) {
        super(manager);
    }

    @Override
    public @NonNull String getId() {
        return "join";
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull ConfigReader params
    ) {
        // 优先 values（现代别名），否则 list（旧键）；仅当值不是 List 时视为 null
        val valuesObject = params.containsKey("values")
            ? params.get("values")
            : params.get("list");
        List<String> values = null;
        if (valuesObject instanceof List<?>) {
            values = new ArrayList<>();
            for (val object : (List<?>) valuesObject) {
                if (object == null) continue;
                values.add(object.toString());
            }
        }
        return handle(
            context,
            values,
            getParsedValue(context, params, "separator"),
            getParsedValue(context, params, "prefix"),
            getParsedValue(context, params, "postfix"),
            getParsedValue(context, params, "limit"),
            getParsedValue(context, params, "truncated"),
            // transform 文本本身不做节点解析，对齐旧 section
            params.getString("transform"),
            getParsedValue(context, params, "shuffled")
        );
    }

    private @Nullable String handle(
        @NonNull ActionContext context,
        @Nullable List<String> list,
        @Nullable String rawSeparator,
        @Nullable String rawPrefix,
        @Nullable String rawPostfix,
        @Nullable String rawLimit,
        @Nullable String truncated,
        @Nullable String rawTransform,
        @Nullable String rawShuffled
    ) {
        // 旧实现：list 为 null 才返回 null；空列表继续拼 prefix+postfix
        if (list == null) return null;

        val separator = rawSeparator == null ? ", " : rawSeparator;
        val prefix = rawPrefix == null ? "" : rawPrefix;
        val postfix = rawPostfix == null ? "" : rawPostfix;

        // 非法 limit → 无限制（对齐 toIntOrNull 失败）
        Integer limit = null;
        if (rawLimit != null && !rawLimit.isEmpty()) {
            val parsedLimit = NumberParser.parseInteger(rawLimit);
            if (parsedLimit != null) {
                if (parsedLimit < 0) {
                    limit = 0;
                } else if (parsedLimit < list.size()) {
                    limit = parsedLimit;
                }
            }
        }

        // toBooleanStrictOrNull：仅精确 "true"/"false"（大小写敏感）
        boolean shuffled = "true".equals(rawShuffled);

        List<String> realList = list;
        if (shuffled) {
            realList = new ArrayList<>(list);
            Collections.shuffle(realList, ThreadLocalRandom.current());
        }

        CompiledScript transform = null;
        if (rawTransform != null && !rawTransform.isEmpty()) {
            transform = compiledScripts.computeIfAbsent(rawTransform, script -> new CompiledScript(
                "function main() {\n" +
                    script + "\n" +
                    "}"
            ));
        }

        val result = new StringBuilder();
        result.append(prefix);
        val length = limit == null ? realList.size() : limit;

        val map = new HashMap<String, Object>();
        if (transform != null) {
            val caster = context.getCaster();
            if (caster instanceof OfflinePlayer) {
                map.put("player", caster);
            }
            map.put("list", realList);
            map.put("vars", (Function<String, String>) string -> this.manager.parseNode(string, context));
        }

        for (int index = 0; index < length; index++) {
            // 先取原始串再 parse，对齐旧「shuffle 原始 → 循环内 parse」
            var element = this.manager.parseNode(realList.get(index), context);
            if (transform != null) {
                map.put("it", element);
                map.put("index", index);
                val transformed = transform.invoke("main", map);
                element = transformed == null ? "" : transformed.toString();
            }
            result.append(element);
            if (index != length - 1 || (limit != null && truncated != null)) {
                result.append(separator);
            }
        }
        if (limit != null && truncated != null) {
            result.append(truncated);
        }
        result.append(postfix);
        return result.toString();
    }
}

File: src/main/kotlin/pers/neige/neigeitems/section/impl/JoinParser.kt
package pers.neige.neigeitems.section.impl

import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import pers.neige.neigeitems.script.CompiledScript
import pers.neige.neigeitems.section.SectionParser
import pers.neige.neigeitems.utils.SectionUtils.parseSection
import java.util.concurrent.ConcurrentHashMap

/**
 * join节点解析器
 */
object JoinParser : SectionParser() {
    override val id: String = "join"

    /**
     * 获取所有用于join节点的已编译的js脚本文件及文本
     */
    val compiledScripts = ConcurrentHashMap<String, CompiledScript>()

    override fun onRequest(
        data: ConfigurationSection,
        cache: HashMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String? {
        return handler(
            cache,
            player,
            sections,
            data.getStringList("list"),
            data.getString("separator"),
            data.getString("prefix"),
            data.getString("postfix"),
            data.getString("limit"),
            data.getString("truncated"),
            data.getString("transform"),
            data.getString("shuffled")
        )
    }

    /**
     * @param cache 解析值缓存
     * @param player 待解析玩家
     * @param sections 节点池
     * @param list 待操作列表
     * @param rawSeparator 分隔符
     * @param rawPrefix 前缀
     * @param rawPostfix 后缀
     * @param rawLimit 长度限制
     * @param rawTruncated 删节符号
     * @param rawTransform 操作函数
     * @return 解析值
     */
    private fun handler(
        cache: HashMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?,
        list: List<String>?,
        rawSeparator: String?,
        rawPrefix: String?,
        rawPostfix: String?,
        rawLimit: String?,
        rawTruncated: String?,
        rawTransform: String?,
        rawShuffled: String?
    ): String? {
        // 如果待操作列表存在, 进行后续操作
        list?.let {
            // 获取分隔符(默认为", ")
            val separator = rawSeparator?.parseSection(cache, player, sections) ?: ", "
            // 获取前缀
            val prefix = rawPrefix?.parseSection(cache, player, sections) ?: ""
            // 获取后缀
            val postfix = rawPostfix?.parseSection(cache, player, sections) ?: ""
            // 获取长度限制
            val limit = rawLimit?.parseSection(cache, player, sections)?.toIntOrNull()?.let {
                when {
                    it >= list.size -> null
                    it < 0 -> 0
                    else -> it
                }
            }
            // 获取删节符号
            val truncated = rawTruncated?.parseSection(cache, player, sections)
            // 获取操作函数
            val transform = rawTransform?.let {
                compiledScripts.computeIfAbsent(it) {
                    CompiledScript("""
                        function main() {
                            $it
                        }""".trimIndent())
                }
            }

            // 开始构建结果
            val result = StringBuilder()
            // 添加前缀
            result.append(prefix)
            // 获取遍历范围
            val length = limit ?: list.size

            // 获取是否乱序
            val shuffled = rawShuffled?.parseSection(cache, player, sections)?.toBooleanStrictOrNull() ?: false
            val realList = if (shuffled) {
                list.shuffled()
            } else {
                list
            }

            // 预定义参数map
            val map = HashMap<String, Any>()
            // 预添加参数
            transform?.let {
                player?.let {
                    // 玩家
                    map["player"] = player
                }
                // 待操作列表
                map["list"] = realList
                // 节点解析函数
                map["vars"] = java.util.function.Function<String, String> { string -> string.parseSection(cache, player, sections) }
            }

            // 遍历列表
            for (index in 0 until length) {
                // 解析元素节点
                var element = realList[index].parseSection(cache, player, sections)
                // 操作元素
                transform?.let {
                    // 待操作元素
                    map["it"] = element
                    // 当前序号
                    map["index"] = index
                    // 操作元素
                    element = transform.invoke("main", map)?.toString() ?: ""
                }
                // 添加元素
                result.append(element)
                // 添加分隔符
                if (index != (length - 1) || (limit != null && truncated != null)) {
                    result.append(separator)
                }
            }
            // 添加删节符号
            if (limit != null && truncated != null) {
                result.append(truncated)
            }
            // 添加后缀
            result.append(postfix)
            // 返回结果
            return result.toString()
        }
        return null
    }
}